### PR TITLE
(Binary Serialization Part 3) Refactor frames

### PR DESCRIPTION
**Pull request by dgant**
#213

I'll be issuing a series of pull requests which cover the full transition to binary serialization.

---

I'm splitting up frame.h and frame.cpp because they're unwieldy, and in anticipation of adding gobs of binary serialization boilerplate that would make them unwield-ier.

@jgehring @ebetica @syhw

### CODE REVIEW

File: include/frame.h [hunk 1]
```diff
@@ -24,506 +24,381 @@ typedef unsigned short uint16_t;
...
+    Frame(const Frame* o);
+    Frame& operator=(Frame& other);
+
+    void swap(Frame& a, Frame& b);
+    void clear();
+    void filter(int32_t x, int32_t y, Frame& o) const;
+    void combine(const Frame& next_frame);
+    bool getCreepAt(uint32_t x, uint32_t y);
+    
+
```
Comment (jgehring) [inline include/frame.h]: Nit: extra newlines

**Comment by dgant** [comment]
@jgehring yeah, should be a no-op diff. Is there a quick way to apply clang-format? Don't seem to have it installed or as part of the container

**Comment by jgehring** [comment]
clang-format-3.8 should be on the cluster. If run from the root of the source tree it will pick up the correct style.

**Comment by dgant** [comment]
@jgehring clang-format-3.8 doesn't seem to like our current config:

```
Error reading .clang-format: Invalid argument
YAML:22:22: error: unknown key 'BreakStringLiterals'
BreakStringLiterals: false
```

If clang-all-the-things is something we want to do regularly it may be worth adding a script to do the recursive clang invocation. I ran it as **find . -name '*.h' -o -name '*.cpp' -o -name '*.cc' -o -name '*.hpp' | xargs clang-format-3.8** -i but that's a bit of a clunker to type each time. Drop into script? 👍 / 👎?

Will just remove the indentation in the meantime.

**Comment by jgehring** [comment]
Oh, I probably ran it on my MacBook only. You can remove this config entry actually, it shouldn't change much.

**Comment by dgant** [comment]
@ebetica @jgehring @syhw With the indentation changes in this PR, and the Keyframe->Frame/Frame->StateUpdate renamings happening in the next PR, is this good to go?

**Comment by syhw** [comment]
Can you provide the command line(s) for a test plan?

**Comment by dgant** [comment]
@syhw The unit tests are in the next PR, and can be run with "make test". I'll update the documentation there to include that. 

The changes up through this PR should just be no-ops. I can demonstrate that they work, but not from within this repo. If we'd prefer to not be merging the intermediate PRs until those tests are ready, that's okay too.

**Comment by jgehring** [comment]
However, note that in general such a script should run `clang-format`, not `clang-format-3.8`. I'll try to have the command without the suffix available on the cluster.

File: BWEnv/CMakeLists.txt [hunk 2]
```diff
@@ -42,7 +42,11 @@ add_library(BWEnvObj OBJECT
   src/user_actions.cc
   src/utils.cc
   src/zmq_server.cc
-  ../replayer/frame.cpp
+  ../replayer/frame_serialization.cpp
+  ../replayer/keyframe.cpp
```
Comment (ebetica) [inline BWEnv/CMakeLists.txt]: Did we decide to leave this as frame.cpp?

File: replayer/frame_diff.cpp [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#include <algorithm>
+
+#include "frame.h"
+
+// TODO: Why is this necessary?
```
Comment (ebetica) [inline replayer/frame_diff.cpp]: What is this comment for? I vaguely recall it always being there but I'm confused so you can probably remove it.

File: BWEnv/CMakeLists.txt (hunk 2 above)
Comment (dgant) [reply BWEnv/CMakeLists.txt]: Yep. Reverted in P4. I'll merge P4 in directly -- P3 doesn't work on its own.

**Comment by dgant** [comment]
This PR has a bunch of changes reverted in Part 4, so I'm closing this and making Part 4 a diff against the development branch

**Pull request closed without merging**